File: src/heavyindustry/world/draw/DrawSpecConstruct.java
package heavyindustry.world.draw;

import heavyindustry.graphics.*;
import arc.graphics.*;
import arc.graphics.g2d.*;
import mindustry.gen.*;
import mindustry.graphics.*;
import mindustry.world.*;
import mindustry.world.draw.*;

import static heavyindustry.util.Utils.*;

/**
 * Display multi-layer textures in sequence according to the progress of the building.
 * @author Wisadell
 */
public class DrawSpecConstruct extends DrawBlock {
    /** Color of Item Surface Construction. */
    public Color constructColor1 = Pal.accent;
    /** The color of the constructed lines. */
    public Color constructColor2 = Pal.accent;

    public int size = 1;

    public TextureRegion[] stageRegions;

    public DrawSpecConstruct() {}

    public DrawSpecConstruct(Color color) {
        constructColor1 = constructColor2 = color;
    }

    public DrawSpecConstruct(Color color1, Color color2) {
        constructColor1 = color1;
        constructColor2 = color2;
    }

    @Override
    public void draw(Building build) {
        int stage = (int) (build.progress() * stageRegions.length);
        float stageProgress = (build.progress() * stageRegions.length) % 1f;

        for (int i = 0; i < stage; i++) {
            Draw.rect(stageRegions[i], build.x, build.y);
        }

        Draw.draw(Layer.blockOver, () -> Drawn.construct(build, stageRegions[stage], constructColor1, constructColor2, 0f, stageProgress, build.warmup() * build.efficiency(), build.totalProgress() * 1.6f));
    }

    @Override
    public void load(Block block) {
        super.load(block);
        if (size < 1) size = 1;
        stageRegions = split(block.name + "-construct", size * 32, 0);
    }
}

File: src/heavyindustry/world/draw/DrawSpecConstruct.kt
package heavyindustry.world.draw

import heavyindustry.graphics.Drawn
import heavyindustry.util.HIUtils
import arc.graphics.Color
import arc.graphics.g2d.Draw
import arc.graphics.g2d.TextureRegion
import mindustry.gen.Building
import mindustry.graphics.Layer
import mindustry.graphics.Pal
import mindustry.world.Block
import mindustry.world.draw.DrawBlock

/** Display multi-layer textures in sequence according to the progress of the building. */
open class DrawSpecConstruct : DrawBlock() {
    /** Sprites layers. */
    @JvmField var stages = 3
    /** Color of Item Surface Construction. */
    @JvmField var constructColor1: Color = Pal.accent
    /** The color of the constructed lines. */
    @JvmField var constructColor2: Color = Pal.accent

    lateinit var stageRegions: Array<TextureRegion>

    override fun draw(build: Building) {
        val stage = (build.progress() * stages).toInt()
        val stageProgress = (build.progress() * stages) % 1f

        for (i in 0 until stage) {
            Draw.rect(stageRegions[i], build.x, build.y)
        }

        Draw.draw(Layer.blockOver) {
            Drawn.construct(build, stageRegions[stage], constructColor1, constructColor2, 0f, stageProgress, build.warmup() * build.efficiency(), build.totalProgress() * 1.6f)
        }
    }

    override fun load(block: Block) {
        stageRegions = HIUtils.split(block.name + "-construct", 32, 0)
    }
}